src/nccl_ofi_dmabuf.cpp

```diff
@@ -41,11 +41,75 @@ static bool kernel_version_rdma_dmabuf_ioctl_ok(void)
 	return (maj == 5 && min >= 12) || (maj > 5);
 }
 
+/**
+ * @brief device_supports_dmabuf - Check if the device supports DMA-BUF
+ * @info: Provider info or hints
+ * @is_prov: true if info is provider, false if hints
+ *
+ * Determines if the device supports DMA-BUF based on its device ID
+ * When passed hints (is_prov = false), it first
+ * gets the actual provider info.
+ *
+ * Return: true if DMA-BUF is supported, false otherwise
+ */
+static bool device_supports_dmabuf(const struct fi_info *info, bool is_prov)
+{
+	struct fi_info *providers = NULL;
+	char *endptr;
+	uint32_t device_id;
+	bool ret = false;
+
+	if (!info) {
+		NCCL_OFI_TRACE(NCCL_INIT | NCCL_NET,
+			       "Invalid input: info is NULL");
+		goto error;
+	}
+
+	/* If hints provided, get actual provider info */
+	if (!is_prov) {
+		int rc = fi_getinfo(FI_VERSION(1, 18), NULL, NULL, 0, info, &providers);
+		if (rc != 0 || !providers) {
+			NCCL_OFI_TRACE(NCCL_INIT | NCCL_NET,
+				       "Failed to get provider info from hints");
+			goto error;
+		}
+		info = providers;
+	}
+
+	/* Check for valid device attributes */
+	if (!info->nic || !info->nic->device_attr || !info->nic->device_attr->device_id) {
+		NCCL_OFI_TRACE(NCCL_INIT | NCCL_NET,
+			       "Invalid device attributes");
+		goto error;
```

Comment:
We should have already queried and filtered for providers before this decision is made. Why do we need to call fi_getinfo() again here?